### ISSUE TITLE
Possible memory leak

### ISSUE THREAD

**Issue by skocibusic** [ngageoint/geopackage-java#43]
Please fill out as much known and relevant information as possible.

#### Version Information:

  * GeoPackage Java Version: 3.0.2
  * GeoPackage Java Source: Maven Central
  * IDE Name & Version: IntelliJ 2018.1
  * Maven Version: -
  * Java Version: $ java -version
java version "1.8.0_181"
Java(TM) SE Runtime Environment (build 1.8.0_181-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.181-b13, mixed mode)
  * Platform & OS: Windows 10
  * Other Relevant Libraries:

#### Expected Results:

  * What did you expect to happen?
When executing my test (copied below), which is opening and closing GeoPackages, I expected to only have a small foot print in the heap space.

#### Observed Results:

  * What happened instead? Memory consumption continuously increased until JVM exits with GC overhead limit exceeded. Checking the VM with VisualVM two ormlite classes consume nearly half of the heap space:
  ** com.j256.ormlite.field.DatabaseFieldConfig; 107.903.760 bytes (26.4%); 899.198 instances (14.2%)
  ** com.j256.ormlite.field.FieldType; 86.323.008 bytes (21.1%); 899.198 instances (14.2%)
  * How often does this occur? Every time.

#### Output:

  * Any logs, errors, or output messages?
```
mil.nga.geopackage.GeoPackageException: Failed to create SpatialReferenceSystem dao

	at mil.nga.geopackage.factory.GeoPackageCoreImpl.createDao(GeoPackageCoreImpl.java:938)
	at mil.nga.geopackage.factory.GeoPackageCoreImpl.getSpatialReferenceSystemDao(GeoPackageCoreImpl.java:373)
	at mil.nga.geopackage.validate.GeoPackageValidate.hasMinimumTables(GeoPackageValidate.java:61)
	at mil.nga.geopackage.validate.GeoPackageValidate.validateMinimumTables(GeoPackageValidate.java:79)
	at mil.nga.geopackage.manager.GeoPackageManager.open(GeoPackageManager.java:116)
	at mil.nga.geopackage.manager.GeoPackageManager.open(GeoPackageManager.java:82)
	at test.LoadTest.loadTest(LoadTest.java:28)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
Caused by: java.sql.SQLException: Could not call the constructor in class class mil.nga.geopackage.core.srs.SpatialReferenceSystemDao
	at com.j256.ormlite.misc.SqlExceptionUtil.create(SqlExceptionUtil.java:27)
	at com.j256.ormlite.dao.DaoManager.createDao(DaoManager.java:95)
	at mil.nga.geopackage.factory.GeoPackageCoreImpl.createDao(GeoPackageCoreImpl.java:935)
	... 28 more
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at com.j256.ormlite.dao.DaoManager.createDao(DaoManager.java:92)
	... 29 more
Caused by: java.sql.SQLException: Could not call the constructor in class class mil.nga.geopackage.core.contents.ContentsDao
	at com.j256.ormlite.misc.SqlExceptionUtil.create(SqlExceptionUtil.java:27)
	at com.j256.ormlite.dao.DaoManager.createDao(DaoManager.java:95)
	at com.j256.ormlite.field.FieldType.configDaoInformation(FieldType.java:392)
	at com.j256.ormlite.dao.BaseDaoImpl.initialize(BaseDaoImpl.java:206)
	at com.j256.ormlite.dao.BaseDaoImpl.<init>(BaseDaoImpl.java:133)
	at com.j256.ormlite.dao.BaseDaoImpl.<init>(BaseDaoImpl.java:112)
	at mil.nga.geopackage.core.srs.SpatialReferenceSystemDao.<init>(SpatialReferenceSystemDao.java:67)
	... 34 more
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at com.j256.ormlite.dao.DaoManager.createDao(DaoManager.java:92)
	... 39 more
Caused by: java.lang.OutOfMemoryError: GC overhead limit exceeded
	at java.util.LinkedHashMap.newNode(LinkedHashMap.java:256)
	at java.util.HashMap.putVal(HashMap.java:631)
	at java.util.HashMap.put(HashMap.java:612)
	at sun.reflect.annotation.AnnotationParser.parseAnnotations2(AnnotationParser.java:124)
	at sun.reflect.annotation.AnnotationParser.parseAnnotations(AnnotationParser.java:72)
	at java.lang.reflect.Field.declaredAnnotations(Field.java:1150)
	at java.lang.reflect.Field.declaredAnnotations(Field.java:1148)
	at java.lang.reflect.Field.getAnnotation(Field.java:1120)
	at com.j256.ormlite.field.DatabaseFieldConfig.fromField(DatabaseFieldConfig.java:515)
	at com.j256.ormlite.field.FieldType.createFieldType(FieldType.java:975)
	at com.j256.ormlite.table.DatabaseTableConfig.extractFieldTypes(DatabaseTableConfig.java:241)
	at com.j256.ormlite.table.DatabaseTableConfig.fromClass(DatabaseTableConfig.java:176)
	at com.j256.ormlite.table.TableInfo.<init>(TableInfo.java:55)
	at com.j256.ormlite.dao.BaseDaoImpl.initialize(BaseDaoImpl.java:156)
	at com.j256.ormlite.dao.BaseDaoImpl.<init>(BaseDaoImpl.java:133)
	at com.j256.ormlite.dao.BaseDaoImpl.<init>(BaseDaoImpl.java:112)
	at mil.nga.geopackage.core.contents.ContentsDao.<init>(ContentsDao.java:63)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at com.j256.ormlite.dao.DaoManager.createDao(DaoManager.java:92)
	at com.j256.ormlite.field.FieldType.configDaoInformation(FieldType.java:392)
	at com.j256.ormlite.dao.BaseDaoImpl.initialize(BaseDaoImpl.java:206)
	at com.j256.ormlite.dao.BaseDaoImpl.<init>(BaseDaoImpl.java:133)
	at com.j256.ormlite.dao.BaseDaoImpl.<init>(BaseDaoImpl.java:112)
	at mil.nga.geopackage.core.srs.SpatialReferenceSystemDao.<init>(SpatialReferenceSystemDao.java:67)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at com.j256.ormlite.dao.DaoManager.createDao(DaoManager.java:92)```
#### Steps to Reproduce:

  1. Execute the test copied below.

#### Relevant Code:

```java
package test;

import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Arrays;
import java.util.List;
import java.util.Random;

import org.junit.Test;

import mil.nga.geopackage.GeoPackage;
import mil.nga.geopackage.manager.GeoPackageManager;

public class LoadTest {
    private static final Path BASE_DIR = Paths.get("D:\\Share\\GPKG");

    private static final List<String> GPKG =
            Arrays.asList([...]);

    private static final Random RNG = new Random();

    @Test
    public void loadTest() {
        for (int i = 0; i < 100_000; i++) {
            GeoPackage geoPackage =
                    GeoPackageManager.open(BASE_DIR.resolve(randomGeoPackage()).toFile());
            geoPackage.close();
        }
    }

    private String randomGeoPackage() {
        return GPKG.get(RNG.nextInt(GPKG.size()));
    }
}
```

#### Test Files:

  * Links to any files needed for testing? Unfortunately I cannot provide any GeoPackages for testing this issue.

#### Additional Information:

I added two screenshot, both from the same execution of the test.
Maybe I oversaw something important, but I think beside opening and closing a GeoPackage there are no further steps necessary. Please correct me if I am wrong.
And feel free to ask questions or request further details.

Cheers,
Marko Skocibusic

![image1](https://user-images.githubusercontent.com/32102517/46084562-3a32b880-c1a4-11e8-994a-414db5e8bd5f.PNG)
![image2](https://user-images.githubusercontent.com/32102517/46084563-3a32b880-c1a4-11e8-8b34-1f8d2f752d52.PNG)

**Comment by bosborn** [comment]
Thanks.  I was able to reproduce the memory leak with other GeoPackages.

Try adding this call directly after the `geoPackage.close()` call:
`DaoManager.clearCache();`

I need to investigate a little more, but it looks like new connection sources cause a new set of retained memory in the Dao Manager each time a GeoPackage is opened.  If this is the fix, I'll add the call as part of the GeoPackage close and release it as part of 3.1.0.

I will be back next week and investigate further, but let me know if that resolves the memory leak on your side.

**Comment by skocibusic** [comment]
Hi Brian,

clearing the cache of the DaoManager does the trick. It fixes my test and is working in my production code. I will probably not call it everytime, but only once in a while. 

Thank you very much for your fast feedback.

Cheers,
Marko

**Comment by bosborn** [comment]
Great, we can leave the issue open until a fix is released.

I'll have to see if there is way to clear the cache for just the closed GeoPackage (for when multiple are open).  I didn't check yet to see what happens to a GeoPackage opened before the cache clear and used after.

**Comment by bosborn** [comment]
Upon closing a GeoPackage, [added a call to clear the DaoManager from known and initialized ConnectionSource DAO types](https://github.com/ngageoint/geopackage-core-java/commit/0872e17308a5b43827992814fbcd48e08d6419cf).  Will release as part of 3.1.0.

Submitted an ORMLite issue and pull request to see if this can be improved upon in the future.

- ConnectionSource DaoManager Memory Cleanup - j256/ormlite-core#150
- unregister all DAOs by a connection source to clear memory - j256/ormlite-core#149

**Comment by bosborn** [comment]
Fixed in [3.1.0](https://github.com/ngageoint/geopackage-java/releases/tag/3.1.0)
